cluster_management/src/main/java/com/pinterest/rocksplicator/Participant.java

```diff
@@ -77,10 +80,13 @@
   private static final String handoffEventHistoryzkSvr = "handoffEventHistoryzkSvr";
   private static final String handoffEventHistoryConfigPath = "handoffEventHistoryConfigPath";
   private static final String handoffEventHistoryConfigType = "handoffEventHistoryConfigType";
+  private static final String  handoffClientEventHistoryJsonShardMapPath = "handoffClientEventHistoryJsonShardMapPath";
 
   private static HelixManager helixManager = null;
   private static LeaderEventsLogger staticParticipantLeaderEventsLogger = null;
   private static LeaderEventsLogger staticSpectatorLeaderEventsLogger = null;
+  private static LeaderEventsLogger staticClientLeaderEventsLogger = null;
+  private static ClientShardMapLeaderEventLoggerDriver staticClientShardMapLeaderEventLoggerDriver = null;
```

Comment:
What is this "Client" vs the Participant and Spectator ?

Reply:
In helix, usually the client is a Spectator that directly listens to helix. In case of Rocksplicator, we only have one Spectator and it is Spectator's responsibility to distribute the shard_map to all interesting parties. This way the client (e.g.. is a separate Root Layer) doesn't depend on helix. So in case of Rocksplicator, The root layer is the client which is interested in updated shard_map information for the purpose of routing requests to specific participant nodes. Participant is instance that actually holds some partitions (shards) and hence some state of the system (e.g.. storage node). Spectator is a translater. Now, in order  to log events from Client's perspective we need to watch the shard_map file that some Spectator has published for given cluster. Since we can't put this in the Root layer (which doesn't exist in rocksplicator itself and that we are not looking for all the clients observations)... we mimick this part of Client in Spectator. Since Spectator itself can run besides Participant, wherever Spectator runs, we also run a partial Client side mimicking the observer of Shard_map to log the delays introduced by in-between systems after Spectator posts the shard_map. This is important to profile, since this can be external system with not a well defined SLA or may not be suitable for the particular use case of distribution of latest shard_map to all interest parties.